### PR TITLE
fix: listing approval changes

### PR DESCRIPTION
**Pull request by ludtkemorgan**
# Pull Request Template

## Issue Overview

This PR addresses #issue

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description

This pulls in the listing approval backend changes from the old backend into the new one. This adds all of the logic to know when to set the new statuses along with creating and send the new emails when applicable.

## How Can This Be Tested/Reviewed?

Unit and e2e tests have been added. But it is recommended to simulate the listing flow using the api endpoints.

The following steps can be done on http://localhost:3101/api after running a reseed of the new backend

Create an admin user - `/user/invite`(change the name and email to the appropriate values along with the jurisdiction id to the one in your prisma db locally
```
{"firstName":"morgan","lastName":"ludtke","email":"morgan.ludtke@exygy.com","roles":{"isAdmin":true,"isPartner":false,"isJurisdictionalAdmin":false},"leasingAgentInListings":[],"jurisdictions":[{"id":"92eb6022-368a-43e9-aa75-4d8430d1266a"}],"agreedToTermsOfService":true}
```

Create a new listing - /listings (change the jurisdictions to the one in your local db)
```
{
  "applicationDueDate": null,
  "applicationFee": null,
  "applicationMethods": [],
  "applicationOpenDate": "2023-10-24T17:55:11.054Z",
  "applicationOrganization": null,
  "applicationPickUpAddress": null,
  "applicationPickUpAddressOfficeHours": null,
  "applicationMailingAddress": null,
  "applicationDropOffAddress": null,
  "applicationDropOffAddressOfficeHours": null,
  "assets": [],
  "buildingSelectionCriteria": null,
  "criteriaAttachType": null,
  "jurisdictions": {
    "name": "Bloomington",
    "id": "92eb6022-368a-43e9-aa75-4d8430d1266a"
  },
  "costsNotIncluded": null,
  "creditHistory": null,
  "criminalBackground": null,
  "depositMax": "0",
  "depositMin": "0",
  "depositHelperText": "or one month's rent may be higher for lower credit scores",
  "disableUnitsAccordion": false,
  "displayWaitlistSize": false,
  "events": [],
  "images": [
    {
      "ordinal": 0,
      "image": { "fileId": "dev/image_6_tffp8w", "label": "cloudinaryBuilding" }
    }
  ],
  "listingFeatures": [],
  "listingUtilities": [],
  "features": {
    "wheelchairRamp": false,
    "elevator": false,
    "serviceAnimalsAllowed": false,
    "accessibleParking": false,
    "parkingOnSite": false,
    "inUnitWasherDryer": false,
    "laundryInBuilding": false,
    "barrierFreeEntrance": false,
    "rollInShower": false,
    "grabBars": false,
    "heatingInUnit": false,
    "acInUnit": false,
    "hearing": false,
    "mobility": false,
    "visual": false
  },
  "utilities": {
    "water": false,
    "gas": false,
    "trash": false,
    "sewer": false,
    "electricity": false,
    "cable": false,
    "phone": false,
    "internet": false
  },
  "leasingAgentAddress": null,
  "leasingAgentEmail": null,
  "leasingAgentName": null,
  "leasingAgentOfficeHours": null,
  "leasingAgentPhone": null,
  "leasingAgentTitle": null,
  "name": "Test listing",
  "postMarkDate": null,
  "postmarkedApplicationsReceivedByDate": null,
  "listingMultiselectQuestions": [],
  "programRules": null,
  "rentalAssistance": "[Bay Area seed] Housing Choice Vouchers, Section 8 and other valid rental assistance programs will be considered for this property. In the case of a valid rental subsidy, the required minimum income will be based on the portion of the rent that the tenant pays after use of the subsidy.",
  "rentalHistory": null,
  "requiredDocuments": null,
  "status": "pending",
  "waitlistCurrentSize": null,
  "waitlistMaxSize": null,
  "isWaitlistOpen": false,
  "waitlistOpenSpots": null,
  "whatToExpect": "Applicants will be contacted by the property agent in rank order until vacancies are filled. All of the information that you have provided will be verified and your eligibility confirmed. Your application will be removed from the waitlist if you have made any fraudulent statements. If we cannot verify a housing preference that you have claimed, you will not receive the preference but will not be otherwise penalized. Should your application be chosen, be prepared to fill out a more detailed application and provide required supporting documents.",
  "units": [],
  "accessibility": null,
  "amenities": null,
  "buildingAddress": null,
  "buildingTotalUnits": 0,
  "developer": "Morgan",
  "householdSizeMax": 0,
  "householdSizeMin": 0,
  "neighborhood": null,
  "petPolicy": null,
  "smokingPolicy": null,
  "unitsAvailable": 0,
  "unitAmenities": null,
  "servicesOffered": null,
  "yearBuilt": null,
  "showWaitlist": false,
  "reviewOrderType": "firstComeFirstServe",
  "unitsSummary": [],
  "unitsSummarized": {
    "unitTypes": [],
    "priorityTypes": [],
    "amiPercentages": [],
    "byUnitTypeAndRent": [],
    "byUnitType": [],
    "byAMI": [],
    "hmi": { "columns": [], "rows": [] }
  },
  "reservedCommunityType": null,
  "dueDateQuestion": null,
  "mapPinPosition": "automatic",
  "reservedCommunityDescription": null,
  "listingAvailabilityQuestion": "availableUnits",
  "specialNotes": null,
  "reviewOrderQuestion": null,
  "waitlistOpenQuestion": "no",
  "digitalApplicationChoice": null,
  "paperApplicationChoice": null,
  "referralOpportunityChoice": null,
  "canApplicationsBeMailedIn": null,
  "canPaperApplicationsBePickedUp": null,
  "canApplicationsBeDroppedOff": null,
  "arePostmarksConsidered": null,
  "additionalApplicationSubmissionNotes": null,
  "applicationDueTimeField": { "hours": "05", "minutes": "00", "period": "pm" },
  "applicationDropOffAddressType": null,
  "applicationPickUpAddressType": null,
  "applicationMailingAddressType": null,
  "digitalApplication": null,
  "commonDigitalApplication": false,
  "paperApplication": null,
  "referralOpportunity": null,
  "customMapPin": false,
  "listingEvents": []
}

````

Create a partner user tied to the listing `/user/invite` (change the listing to the one created above and also appropriate name/email)
```
{
  "lastName": "ludtke",
"firstName": "morgan",
  "listings": [
    {
      "id": "e09952ed-5b30-424d-abcd-71848b569f0a",
      "name": "string"
    }
  ],
  "userRoles": {
    "isAdmin": false,
    "isJurisdictionalAdmin": false,
    "isPartner": true
  },
  "language": "en",
  "jurisdictions": [
    {
    "name": "Bloomington",
    "id": "92eb6022-368a-43e9-aa75-4d8430d1266a"
    }
  ],
  "email": "morgan.ludtke+1@exygy.com"
}
```

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [ ] I have performed a self-review of my own code
- [ ] I have reviewed the changes in a desktop view
- [ ] I have reviewed the changes in a mobile view
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

### CODE REVIEW

**Comment by ColinBuyck** [comment]
Code is looking good, but do you mind adding the step by step instructions for the flows to test?

File: backend_new/test/integration/listing.e2e-spec.ts [hunk 1]
```diff
@@ -572,4 +579,152 @@ describe('Listing Controller Tests', () => {
...
+      expect(mockListingApproved).toBeCalledWith(
+        undefined,
+        { id: listing.id, name: val.name },
+        expect.arrayContaining([partnerUser.email]),
+        jurisdictionA.publicUrl,
+      );
+    });
+
+    // Skipping while trying to figure out req.user
+    it.skip('update status to changes requested and notify appropriate users', async () => {
```
Comment (YazeedLoonat) [inline backend_new/test/integration/listing.e2e-spec.ts:703]: should this still be skipped?

File: backend_new/src/services/email.service.ts [hunk 2]
```diff
@@ -340,4 +329,75 @@ export class EmailService {
...
+        jurisdiction.emailFromAddress,
+        this.polyglot.t('requestApproval.header'),
+        this.template('request-approval')({
+          appOptions: { listingName: listingInfo.name },
+          appUrl: appUrl,
+          listingUrl: `${appUrl}/listings/${listingInfo.id}`,
+        }),
+      );
+    } catch (err) {
+      throw new HttpException('email failed', 500);
```
Comment (YazeedLoonat) [inline backend_new/src/services/email.service.ts:353]: do you think we should add a log here so we can know why this failed?

File: backend_new/src/services/email.service.ts [hunk 3]
```diff
@@ -340,4 +329,75 @@ export class EmailService {
...
+        jurisdiction.emailFromAddress,
+        this.polyglot.t('changesRequested.header'),
+        this.template('changes-requested')({
+          appOptions: { listingName: listingInfo.name },
+          appUrl: appUrl,
+          listingUrl: `${appUrl}/listings/${listingInfo.id}`,
+        }),
+      );
+    } catch (err) {
+      throw new HttpException('email failed', 500);
```
Comment (YazeedLoonat) [inline backend_new/src/services/email.service.ts:377]: same as above

File: backend_new/src/services/email.service.ts [hunk 4]
```diff
@@ -340,4 +329,75 @@ export class EmailService {
...
+        emails,
+        jurisdiction.emailFromAddress,
+        this.polyglot.t('listingApproved.header'),
+        this.template('listing-approved')({
+          appOptions: { listingName: listingInfo.name },
+          listingUrl: `${publicUrl}/listing/${listingInfo.id}`,
+        }),
+      );
+    } catch (err) {
+      throw new HttpException('email failed', 500);
```
Comment (YazeedLoonat) [inline backend_new/src/services/email.service.ts:400]: same

**Review by YazeedLoonat — CHANGES_REQUESTED** [review]
Hey Morgan, looks good just a few questions/changes requested

File: backend_new/src/services/email.service.ts (hunk 2 above)
Comment (ludtkemorgan) [reply backend_new/src/services/email.service.ts:354]: I've added a log above this throw

File: backend_new/test/integration/listing.e2e-spec.ts (hunk 1 above)
Comment (ludtkemorgan) [reply backend_new/test/integration/listing.e2e-spec.ts]: Updated the code to accomodate the login flow again. So this is now working and the other tests are actually work now as well

**Comment by ludtkemorgan** [comment]
@YazeedLoonat this is ready to be re-reviewed

**Review by YazeedLoonat — APPROVED** [review]
LGTM

**Comment by ColinBuyck** [comment]
@ludtkemorgan quick bump on my comment above! Should have time to review today if you're able to get to it 🦞

**Review by ColinBuyck — APPROVED** [review]
LGTM 🦖  Note per offline conversations with @ludtkemorgan that more thorough testing will be done on this piece when the partners side is connected

**Pull request merged**